pkg/reconciler/buildrunttlcleanup/buildrun_ttl_cleanup.go

```diff
@@ -0,0 +1,99 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// ReconcileBuildRun reconciles a BuildRun object
+
+type ReconcileBuildRun struct {
+	/* This client, initialized using mgr.Client() above, is a split client
+	   that reads objects from the cache and writes to the apiserver */
+	config *config.Config
+	client client.Client
+}
+
+func NewReconciler(c *config.Config, mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileBuildRun{
+		config: c,
+		client: mgr.GetClient(),
+	}
+}
+
+/* Reconciler makes sure the buildrun adheres to its ttl retention field and deletes it
+   once the ttl limit is hit */
+func (r *ReconcileBuildRun) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	// Set the ctx to be Background, as the top-level context for incoming requests.
+	ctx, cancel := context.WithTimeout(ctx, r.config.CtxTimeOut)
+	defer cancel()
+
+	ctxlog.Debug(ctx, "Start reconciling Buildrun-ttl", namespace, request.Namespace, name, request.Name)
+
+	br := &buildv1alpha1.BuildRun{}
+	err := r.client.Get(ctx, types.NamespacedName{Name: request.Name, Namespace: request.Namespace}, br)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			ctxlog.Debug(ctx, "Finish reconciling buildrun-ttl. Buildrun was not found", namespace, request.Namespace, name, request.Name)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	condition := br.Status.GetCondition(buildv1alpha1.Succeeded)
+	if condition == nil || condition.Status == corev1.ConditionUnknown {
+		return reconcile.Result{}, nil
+	}
+	var ttl *metav1.Duration
+	if condition.Status == corev1.ConditionTrue {
+		if br.Spec.Retention != nil && br.Spec.Retention.TTLAfterSucceeded != nil {
+			ttl = br.Spec.Retention.TTLAfterSucceeded
+		} else if br.Status.BuildSpec != nil && br.Status.BuildSpec.Retention != nil && br.Status.BuildSpec.Retention.TTLAfterSucceeded != nil {
+			ttl = br.Status.BuildSpec.Retention.TTLAfterSucceeded
+		}
+	} else {
+		if br.Spec.Retention != nil && br.Spec.Retention.TTLAfterFailed != nil {
+			ttl = br.Spec.Retention.TTLAfterFailed
+		} else if br.Status.BuildSpec != nil && br.Status.BuildSpec.Retention != nil && br.Status.BuildSpec.Retention.TTLAfterFailed != nil {
+			ttl = br.Status.BuildSpec.Retention.TTLAfterFailed
+		}
+	}
+
+	// check if BuildRun still has a TTL
+	if ttl == nil {
+		return reconcile.Result{}, nil
+	}
+
+	if br.Status.CompletionTime.Add(ttl.Duration).Before(time.Now()) {
+		ctxlog.Info(ctx, "Deleting buildrun as ttl has been reached.", namespace, request.Namespace, name, request.Name)
+		err := r.client.Delete(ctx, br, &client.DeleteOptions{})
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				ctxlog.Debug(ctx, "Error deleting buildRun.", namespace, request.Namespace, name, br.Name, "error", err)
+				return reconcile.Result{}, err
+			}
+			return reconcile.Result{}, nil
```

Comment:
We were wondering whether it would be unnecessary logging? Considering that the buildrun deletion should have been logged by the controller which performed the action? This log could provide information but is there any scenario where that information could be useful?

Reply:
@gabemontero this is the only open item. I agree that the log does not hurt. It will rarely happen because the BuildRun we delete is one that we just fetched. @raghavbhatnagar96 unfortunately is gone already for today. Can we put this in and add the log statement in a follow-on PR? I would like to grab this PR's changes into the CLI repo and see if I can add flags to build and buildrun commands to make the story complete.

Reply:
I'm fine with that @SaschaSchwarze0 

and yeah I did not expect for this to happen frequently .... it is with regard to dealing with the odd intermittent sort of situations

Reply:
@raghavbhatnagar96 just added it @gabemontero ;-)

Reply:
Yup small golint issue, just resolving that.